models/git/commit_status.go

```diff
@@ -240,6 +241,60 @@ func GetLatestCommitStatus(ctx context.Context, repoID int64, sha string, listOp
 	return statuses, count, db.GetEngine(ctx).In("id", ids).Find(&statuses)
 }
 
+// GetLatestCommitStatusForPairs returns all statuses with a unique context for a given list of repo-sha pairs
+func GetLatestCommitStatusForPairs(ctx context.Context, repoIDs []int64, shas []string, listOptions db.ListOptions) (map[int64][]*CommitStatus, error) {
+	if len(repoIDs) != len(shas) {
+		return nil, errors.New("repoIDs and shas must have the same length")
+	}
+
+	type Result struct {
+		ID     int64
+		RepoID int64
+	}
+
+	results := make([]Result, 0)
+
+	sess := db.GetEngine(ctx).Table(&CommitStatus{})
+
+	// Create a disjunction of conditions for each repoID and SHA pair
+	conds := make([]builder.Cond, 0, len(repoIDs))
+	for i := range repoIDs {
+		conds = append(conds, builder.And(builder.Eq{"repo_id": repoIDs[i]}, builder.Eq{"sha": shas[i]}))
+	}
+	sess = sess.Where(builder.Or(conds...)).
+		Select("max( id ) as id, repo_id").
+		GroupBy("context_hash, repo_id").OrderBy("max( id ) desc")
```

Comment:
Furthermore, isn't your only goal the **latest** commit status of any repo?
So, why return a list of them for each repo then?
Is it because you can upload many different commit statuses for a single commit?
Even if yes, it's probably better to aggregate them, and only returning one combined status for each repo as that's what we want here.

Reply:
I was following the approach in https://github.com/go-gitea/gitea/blob/365bb77a5497d492836bf823cf780c823db27e8c/routers/web/repo/view.go#L836-L841

I will try to do what you suggest

Reply:
Nevermind, I think we still need it because each repo can have more than one `context_hash` (like we have ~15 at the moment)
![image](https://github.com/go-gitea/gitea/assets/20454870/a95194b2-80b2-45f9-8bb4-16917a530e18)

Reply:
See the last sentence which was exactly intended for this:
> Even if yes, it's probably better to aggregate them, and only returning one combined status for each repo as that's what we want here.

After all, there's already an API route that does the same, so it should be fairly easy.

Reply:
Do you mean calling `git_model.CalcCommitStatus` inside `GetLatestCommitStatusForPairs`?

Reply:
@delvh I'm not sure what you mean

Reply:
> If I see this correctly, your code currently works because every subsequent insert gets a higher ID, right? I would've guessed that you want to sort by `created_unix DESC`.

It's better to use id, created_unix maybe equal but id will not.

Reply:
The problem here is the `branch`. The result should only for default branch of this repository. But maybe it hasn't been stored into `commit_status`.

Commit status of a pull request branch or a non-default branch may not be meaningful for the default branch.

I remember there is a feature request about the badge of CI, maybe it's related. Maybe this issue could depend on that one.